mypy/join.py

```diff
@@ -17,6 +17,11 @@
 def join_simple(declaration: Type, s: Type, t: Type) -> Type:
     """Return a simple least upper bound given the declared type."""
 
+    if (s.can_be_true, s.can_be_false) != (t.can_be_true, t.can_be_false):
```

Comment:
`join_types` also reuses existing values in some cases.  I'd expect the same problem to exist, but just be pretty hard to actually repro.